File: src/systems/Traps.js
import * as THREE from 'three';
import { makeTrapPanel, makeTrapEmitter } from '../world/Props.js';
import { Interactable } from './Interaction.js';
import { MAP } from '../world/MapData.js';
import { Audio } from '../core/AudioEngine.js';
import { clamp, rand, pick } from '../core/Util.js';

const ACTIVE_TIME = 22;
const COOLDOWN = 40;

const KINDS = {
  electric: { dps: 300, label: 'ARC PYLONS', color: 0x66d9ff, stun: true },
  fire: { dps: 240, label: 'FLAME VENTS', color: 0xff7b2e, burn: true },
  turret: { dps: 0, label: 'SENTRY GUN', color: 0xffd24a, shots: true },
};

class Trap extends Interactable {
  constructor(def, system) {
    super({
      id: def.id, x: def.panel.x, y: def.y + 1.2, z: def.panel.z, radius: 2.4, priority: 1,
    });
    this.def = def;
    this.kind = KINDS[def.kind];
    this.system = system;
    this.state = 'ready';
    this.t = 0;

    this.panel = makeTrapPanel(def.kind);
    this.panel.position.set(def.panel.x, def.y + 1.35, def.panel.z);
    this.panel.rotation.y = def.panel.rot;

    this.emitter = makeTrapEmitter(def.kind);
    this.emitter.position.set(def.x, def.y, def.z);
    this.emitter.rotation.y = def.rot;

    this.light = new THREE.PointLight(this.kind.color, 0, 20);
    this.light.position.set(def.x, def.y + 2.2, def.z);

    this.shotT = 0;
  }

  get powered() { return this.system.power.on; }

  prompt() {
    if (!this.powered) return { text: this.kind.label, sub: 'NO POWER', blocked: true };
    if (this.state === 'active') return { text: this.kind.label, sub: 'ACTIVE', blocked: true };
    if (this.state === 'cooldown') {
      return { text: this.kind.label, sub: `RECHARGING ${Math.ceil(this.t)}s`, blocked: true };
    }
    return { text: `ARM ${this.kind.label}`, cost: this.def.cost, sub: `${ACTIVE_TIME}s OF CARNAGE` };
  }

  activate(ctx) {
    if (!this.powered || this.state !== 'ready') return false;
    if (!ctx.player.spend(this.def.cost)) { ctx.deny(); return false; }
    this.state = 'active';
    this.t = ACTIVE_TIME;
    Audio.playAt('trap', { x: this.def.x, y: this.def.y + 1, z: this.def.z }, { kind: this.def.kind, volume: 1 });
    ctx.toast(`${this.kind.label} ARMED`, 'warn');
    this.system.world.nav.setDanger(this.def.area, 3);
    return true;
  }

  update(dt, time, ctx) {
    const face = this.panel.userData.face;
    const on = this.powered;

    if (this.state === 'active') {
      this.t -= dt;
      this.light.intensity = 90 + Math.sin(time * 30) * 40;
      face.material.emissiveIntensity = 2.4;
      this._runEffect(dt, time, ctx);
      if (this.t <= 0) {
        this.state = 'cooldown';
        this.t = COOLDOWN;
        this.system.world.nav.setDanger(this.def.area, 0);
      }
    } else {
      this.light.intensity *= Math.exp(-6 * dt);
      face.material.emissiveIntensity = on
        ? (this.state === 'ready' ? 1.4 + Math.sin(time * 2) * 0.4 : 0.35)
        : 0.05;
      if (this.state === 'cooldown') {
        this.t -= dt;
        if (this.t <= 0) this.state = 'ready';
      }
      this._idleVisual(dt, time);
    }
  }

  _idleVisual(dt, time) {
    if (this.def.kind === 'electric') {
      for (const c of this.emitter.userData.coils || []) {
        c.material.emissiveIntensity = this.state === 'active' ? 3 : 0.5;
      }
    } else if (this.def.kind === 'fire') {
      for (const j of this.emitter.userData.jets || []) j.visible = false;
    } else if (this.emitter.userData.head) {
      this.emitter.userData.head.rotation.y = Math.sin(time * 0.4) * 0.7;
    }
  }

  _inArea(z) {
    const a = this.def.area;
    return z.pos.x > a.x0 && z.pos.x < a.x1 && z.pos.z > a.z0 && z.pos.z < a.z1;
  }

  _runEffect(dt, time, ctx) {
    const targets = ctx.zombies.alive.filter((z) => z.active && z.state !== 6 && this._inArea(z));

    if (this.def.kind === 'electric') {
      for (const c of this.emitter.userData.coils || []) c.material.emissiveIntensity = 3.5;
      if (Math.random() < dt * 26) {
        const a = this.emitter.children[1]?.position ?? new THREE.Vector3();
        ctx.fx.arc(
          { x: this.def.x - 2.6, y: this.def.y + 2.9, z: this.def.z },
          { x: this.def.x + 2.6, y: this.def.y + 2.9, z: this.def.z },
          this.kind.color,
        );
        void a;
      }
      for (const z of targets) {
        const res = z.takeDamage(this.kind.dps * dt, 'torso', null, null);
        z.applySlow(0.45, 0.4);
        if (Math.random() < dt * 4) {
          ctx.fx.arc(
            { x: this.def.x, y: this.def.y + 2.6, z: this.def.z },
            { x: z.pos.x, y: z.pos.y + 1.2, z: z.pos.z }, this.kind.color,
          );
        }
        if (res.killed) ctx.onTrapKill(z, this);
      }
    } else if (this.def.kind === 'fire') {
      for (const j of this.emitter.userData.jets || []) {
        j.visible = true;
        j.scale.y = 1 + Math.sin(time * 24 + j.position.x) * 0.35;
      }
      if (Math.random() < dt * 30) {
        ctx.fx.emberBurst({
          x: this.def.x + rand(-3.5, 3.5), y: this.def.y + rand(0.3, 1.6), z: this.def.z + rand(-0.6, 0.6),
        });
      }
      for (const z of targets) {
        const res = z.takeDamage(this.kind.dps * dt, 'torso', null, null);
        z.applyBurn(90, 3);
        if (res.killed) ctx.onTrapKill(z, this);
      }
    } else {
      // Sentry: pick the closest target and put rounds into it.
      const head = this.emitter.userData.head;
      if (targets.length) {
        const t = targets.reduce((a, b) => (
          Math.hypot(a.pos.x - this.def.x, a.pos.z - this.def.z) < Math.hypot(b.pos.x - this.def.x, b.pos.z - this.def.z) ? a : b
        ));
        const ang = Math.atan2(t.pos.x - this.def.x, t.pos.z - this.def.z) - this.def.rot;
        if (head) head.rotation.y = ang;
        this.shotT -= dt;
        if (this.shotT <= 0) {
          this.shotT = 0.14;
          const from = { x: this.def.x, y: this.def.y + 0.85, z: this.def.z };
          const to = { x: t.pos.x, y: t.pos.y + 1.1, z: t.pos.z };
          ctx.fx.tracer(from, to, 0xffd24a);
          Audio.playAt('shoot', from, { weight: 'light', volume: 0.5 });
          const res = t.takeDamage(320, pick(['head', 'torso', 'torso']), { headMult: 2.2, torsoMult: 1 }, ctx.fx);
          if (res.killed) ctx.onTrapKill(t, this);
        }
      } else if (head) {
        head.rotation.y = Math.sin(time * 1.2) * 0.6;
      }
    }
  }
}

export class TrapSystem {
  constructor(world, scene, interaction, power) {
    this.world = world;
    this.power = power;
    this.traps = [];
    for (const def of MAP.traps) {
      const t = new Trap(def, this);
      scene.add(t.panel, t.emitter, t.light);
      interaction.add(t);
      this.traps.push(t);
    }
  }

  update(dt, time, ctx) {
    for (const t of this.traps) t.update(dt, time, ctx);
  }

  reset() {
    for (const t of this.traps) {
      t.state = 'ready';
      t.t = 0;
      this.world.nav.setDanger(t.def.area, 0);
    }
  }
}

export { clamp };

File: unity/Rotgrid/Assets/Scripts/Systems/Traps.cs
using System.Collections.Generic;
using UnityEngine;

namespace Rotgrid
{
    /// <summary>
    /// An armable hazard: arc pylons, flame vents or a sentry gun. Costs points,
    /// runs on a timer, then cools down.
    /// </summary>
    public class Trap : Interactable
    {
        const float ActiveTime = 22f;
        const float Cooldown = 40f;

        public readonly TrapDef def;
        readonly TrapSystem _system;
        readonly MachineInfo _panel;
        readonly MachineInfo _emitter;
        readonly Light _light;
        readonly string _label;
        readonly float _dps;
        readonly Color _color;

        string _state = "ready";
        float _t, _shotT;

        public Trap(TrapDef d, TrapSystem system, Transform parent)
            : base(d.id, new Vector3(d.panelX, d.y + 1.2f, d.panelZ), 2.4f, 1)
        {
            def = d;
            _system = system;
            _color = Props.TrapColor(d.kind);
            if (d.kind == "electric") { _label = "ARC PYLONS"; _dps = 300f; }
            else if (d.kind == "fire") { _label = "FLAME VENTS"; _dps = 240f; }
            else { _label = "SENTRY GUN"; _dps = 0f; }

            var panelGo = Props.TrapPanel(d.kind, parent);
            panelGo.transform.position = new Vector3(d.panelX, d.y + 1.35f, d.panelZ);
            panelGo.transform.rotation = Quaternion.Euler(0, d.panelRot, 0);
            _panel = panelGo.GetComponent<MachineInfo>();

            var emitterGo = Props.TrapEmitter(d.kind, parent);
            emitterGo.transform.position = new Vector3(d.x, d.y, d.z);
            emitterGo.transform.rotation = Quaternion.Euler(0, d.rot, 0);
            _emitter = emitterGo.GetComponent<MachineInfo>();

            var lightGo = new GameObject("trapLight");
            lightGo.transform.SetParent(parent, false);
            lightGo.transform.position = new Vector3(d.x, d.y + 2.2f, d.z);
            _light = lightGo.AddComponent<Light>();
            _light.type = LightType.Point;
            _light.color = _color;
            _light.range = 20f;
            _light.intensity = 0f;
            _light.shadows = LightShadows.None;
        }

        bool Powered { get { return _system.power.On; } }

        public override Prompt GetPrompt(GameRun run)
        {
            if (!Powered) return Prompt.Blocked(_label, "NO POWER");
            if (_state == "active") return Prompt.Blocked(_label, "ACTIVE");
            if (_state == "cooldown") return Prompt.Blocked(_label, "RECHARGING " + Mathf.Ceil(_t) + "s");
            return Prompt.Buy("ARM " + _label, def.cost, ActiveTime + "s OF CARNAGE");
        }

        public override bool Activate(GameRun run)
        {
            if (!Powered || _state != "ready") return false;
            if (!run.player.Spend(def.cost)) { run.Deny(); return false; }
            _state = "active";
            _t = ActiveTime;
            Game.Audio.PlayAt(def.kind == "fire" ? Sfx.TrapFire : Sfx.TrapElectric,
                new Vector3(def.x, def.y + 1f, def.z), 1f);
            run.Toast(_label + " ARMED", ToastKind.Warn, false);
            _system.world.nav.SetDanger(def.area, 3f);
            return true;
        }

        public void Update(float dt, float time, GameRun run)
        {
            var face = (_panel.glow != null && _panel.glow.Length > 0) ? _panel.glow[0] : null;

            if (_state == "active")
            {
                _t -= dt;
                _light.intensity = 90f + Mathf.Sin(time * 30f) * 40f;
                if (face != null) Art.SetEmission(face.material, _color * 2.4f);
                RunEffect(dt, time, run);
                if (_t <= 0f)
                {
                    _state = "cooldown";
                    _t = Cooldown;
                    _system.world.nav.SetDanger(def.area, 0f);
                    if (def.kind == "fire" && _emitter.extras != null)
                        foreach (var j in _emitter.extras) if (j != null) j.SetActive(false);
                }
            }
            else
            {
                _light.intensity *= Mathf.Exp(-6f * dt);
                if (face != null)
                {
                    float k = Powered ? (_state == "ready" ? 1.4f + Mathf.Sin(time * 2f) * 0.4f : 0.35f) : 0.05f;
                    Art.SetEmission(face.material, _color * k);
                }
                if (_state == "cooldown")
                {
                    _t -= dt;
                    if (_t <= 0f) _state = "ready";
                }
                if (def.kind == "turret" && _emitter.pivot != null)
                    _emitter.pivot.localRotation = Quaternion.Euler(0, Mathf.Sin(time * 1.2f) * 34f, 0);
            }
        }

        bool InArea(Zombie z)
        {
            return z.pos.x > def.area.x0 && z.pos.x < def.area.x1
                && z.pos.z > def.area.z0 && z.pos.z < def.area.z1;
        }

        void RunEffect(float dt, float time, GameRun run)
        {
            var targets = new List<Zombie>();
            foreach (var z in run.zombies.alive)
                if (z.active && z.state != ZState.Dying && InArea(z)) targets.Add(z);

            if (def.kind == "electric")
            {
                if (_emitter.glow != null)
                    foreach (var c in _emitter.glow) if (c != null) Art.SetEmission(c.material, _color * 3.5f);
                if (Random.value < dt * 26f)
                    run.fx.Arc(new Vector3(def.x - 2.6f, def.y + 2.9f, def.z),
                               new Vector3(def.x + 2.6f, def.y + 2.9f, def.z), _color);
                foreach (var z in targets)
                {
                    var res = z.TakeDamage(_dps * dt, BodyPart.Torso, null, null, Vector3.zero);
                    z.ApplySlow(0.45f, 0.4f);
                    if (Random.value < dt * 4f)
                        run.fx.Arc(new Vector3(def.x, def.y + 2.6f, def.z), z.pos + Vector3.up * 1.2f, _color);
                    if (res.killed) run.OnTrapKill(z);
                }
            }
            else if (def.kind == "fire")
            {
                if (_emitter.extras != null)
                {
                    foreach (var j in _emitter.extras)
                    {
                        if (j == null) continue;
                        j.SetActive(true);
                        var s = j.transform.localScale;
                        j.transform.localScale = new Vector3(s.x, 0.3f * (1f + Mathf.Sin(time * 24f + j.transform.localPosition.x) * 0.35f), s.z);
                    }
                }
                if (Random.value < dt * 30f)
                    run.fx.EmberBurst(new Vector3(def.x + Util.Rand(-3.5f, 3.5f), def.y + Util.Rand(0.3f, 1.6f),
                                                  def.z + Util.Rand(-0.6f, 0.6f)));
                foreach (var z in targets)
                {
                    var res = z.TakeDamage(_dps * dt, BodyPart.Torso, null, null, Vector3.zero);
                    z.ApplyBurn(90f, 3f);
                    if (res.killed) run.OnTrapKill(z);
                }
            }
            else
            {
                // Sentry: pick the closest target and put rounds into it.
                if (targets.Count > 0)
                {
                    Zombie best = targets[0];
                    float bestD = float.MaxValue;
                    foreach (var z in targets)
                    {
                        float d = (z.pos.x - def.x) * (z.pos.x - def.x) + (z.pos.z - def.z) * (z.pos.z - def.z);
                        if (d < bestD) { bestD = d; best = z; }
                    }
                    if (_emitter.pivot != null)
                    {
                        float ang = Mathf.Atan2(best.pos.x - def.x, best.pos.z - def.z) * Mathf.Rad2Deg - def.rot;
                        _emitter.pivot.localRotation = Quaternion.Euler(0, ang, 0);
                    }
                    _shotT -= dt;
                    if (_shotT <= 0f)
                    {
                        _shotT = 0.14f;
                        var from = new Vector3(def.x, def.y + 0.85f, def.z);
                        run.fx.Tracer(from, best.pos + Vector3.up * 1.1f, Util.Hex(0xffd24a));
                        Game.Audio.PlayAt(Sfx.ShootLight, from, 0.5f);
                        var part = Random.value < 0.34f ? BodyPart.Head : BodyPart.Torso;
                        var res = best.TakeDamage(320f, part, null, run.fx, Vector3.zero);
                        if (res.killed) run.OnTrapKill(best);
                    }
                }
                else if (_emitter.pivot != null)
                    _emitter.pivot.localRotation = Quaternion.Euler(0, Mathf.Sin(time * 1.2f) * 34f, 0);
            }
        }

        public void ResetState()
        {
            _state = "ready";
            _t = 0f;
            _system.world.nav.SetDanger(def.area, 0f);
        }
    }

    public class TrapSystem
    {
        public readonly WorldBuilder world;
        public readonly PowerSystem power;
        public readonly List<Trap> traps = new List<Trap>();

        public TrapSystem(WorldBuilder w, PowerSystem p, InteractionSystem interaction, Transform parent)
        {
            world = w;
            power = p;
            foreach (var d in MapData.Traps)
            {
                var t = new Trap(d, this, parent);
                interaction.Add(t);
                traps.Add(t);
            }
        }

        public void Update(float dt, float time, GameRun run)
        {
            foreach (var t in traps) t.Update(dt, time, run);
        }
    }
}
